Add support for mesh ingress

fosterseth opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
API changes to support mesh ingress (in-cluster hop node).

Notable changes:
- adds ReceptorAddress model and related endpoints to the API
- adds managed field to instance
- InstanceLink peering is now from Instance to ReceptorAddress (before it was Instance to Instance)
- UI changes to support the above peering changes, including a new Endpoints tab on the Instance detail page

 
<!---
If you are fixing an existing issue, please include "related #nnn" in your
commit message and your description; but you should still explain what
the change does.
-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - New or Enhanced Feature

##### COMPONENT NAME
<!--- Name of the module/plugin/module/task -->
 - API
 - UI
 - Collection

fosterseth committed:
Add support for inbound hop nodes

TheRealHaoLiu committed:
Connect from controlplane node to mesh ingress

fosterseth committed:
Add ReceptorAddress to root urls

- Add database contraints to make sure addresses
are unique
If port is defined:
address, port, protocol, websocket_path are unique together

if port is not defined:
address, protocol, websocket_path are unique together

- Allow deleting address via API
- Add ReceptorAddressAccess to determine permissions
- awx-manage add_receptor_address returns changed: True
if successful

fosterseth committed:
Update receptor conf when address changes

Add post save and post delete hooks to
call write_receptor_config when
a receptor address is added / removed.

Add peers_from_control_nodes to
provision_instance

Signed-off-by: Seth Foster <fosterbseth@gmail.com>

fosterseth committed:
Add peers_from_control_nodes to ReceptorAddress

- write_receptor_config peers to ReceptorAddress entries
that have peers_from_control_nodes enabled

- peers_from_control_nodes and listener_port removed from Instance model

- peers_from_control_nodes added to ReceptorAddress model

- ReceptorAddress is now unique by address and protocol combination

- Write receptor config task is dispatched upon ReceptorAddress creation
or deletion, and when control node is first created

- InstanceLinkSerializer adds a target_address field and has logic
to grab the instance hostname associated with the peered ReceptorAddress

Signed-off-by: Seth Foster <fosterbseth@gmail.com>

fosterseth committed:
Add install bundle support

group_vars all.yaml changes:
- peer entry has two fields, address and port
- receptor_port is inferred from the first
receptor_address entry that uses protocol tcp

other changes:
ActivityStream now records when receptor_addresses
are peered to

Signed-off-by: Seth Foster <fosterbseth@gmail.com>

fosterseth committed:
Register_peers support for receptor_addresses

register_peers has inputs:

source: source instance
peers: list of instances the source should peer to

InstanceLink "target" is now expected to be a ReceptorAddress

For each peer, we can just use the first receptor address. If
multiple receptor addresses exist, throw a command error.

Currently this command is only used on VM-deployments, where
there is only a single receptor address per instance, so this
should work fine.

Other changes:
drop listener_port field from Instance. Listener port is now just
"port" on ReceptorAddress

Signed-off-by: Seth Foster <fosterbseth@gmail.com>

TheRealHaoLiu committed:
Temp change to aid dev

Temporarily switch to feature branch for receptor-collection in instance_install_bundle to make dev easier

AlanCoding commented on awx/main/migrations/0188_inbound_hop_nodes.py:
```diff
@@ -0,0 +1,68 @@
+            name='instancelink',
+            unique_together=set(),
+        ),
+        migrations.RemoveField(
```
Before this data is removed, I take it that a data migration needs to be added to create ReceptorAddress objects for old instances. It looks like that will also require re-ordering these steps.

fosterseth replied:
yep this is this planned work item we still need to implement

AlanCoding commented on awx/main/models/ha.py:
```diff
@@ -62,8 +63,11 @@ def has_policy_changes(self):
+        ordering = ("id",)
+
+    source = models.ForeignKey('Instance', on_delete=models.CASCADE)
+    target = models.ForeignKey('ReceptorAddress', on_delete=models.CASCADE)
```
Please can we get `help_text` added to these? I don't feel like the field names here ever were self-explanatory, but with the model for the target changing, we go far beyond any ability to naturally decipher it. This is particularly explosive because the value is an integer that used to represent the id of one model, and now is the id of a different model, and may be the same, or different, as before the migration, arbitrarily.

AlanCoding commented on awx/main/models/receptor_address.py:
```diff
@@ -0,0 +1,56 @@
+        elif self.protocol in ['ws', 'wss']:
+            return 'ws-peer'
+        else:
+            return None
```
I don't understand why you would do this method as opposed to making `protocol` a choice field with the choices of `['tcp-peer', 'ws-peer']`

fosterseth replied:
I've made `protocol` a choice field with options "tcp", "ws" and "wss"

in the real world "tcp" is a protocol, but "tcp-peer" is more of a receptor config item name.

AlanCoding commented on awx/main/models/ha.py:
```diff
@@ -499,6 +492,28 @@ def schedule_write_receptor_config(broadcast=True):
+    if address.peers_from_control_nodes:
+        address.peers_from.add(*control_instances)
+    else:
+        address.peers_from.remove(*control_instances)
```
What about when a `ReceptorAddress` is saved, where that is the address for a control node?

We have specific rules are which control nodes are connected to which, and it is not bi-directional. I don't know what the value of `peers_from_control_nodes` will be for addresses in the control plane, and I don't know if it is relevant. But this would seem to fire for all saves of all receptor addresses, which would happen for the control nodes simply due to the fact that they have an associated receptor address, and it looks like that right away puts the links in the wrong state.

fosterseth replied:
I've since cleaned this up a bit. The logic is as follows:

if address has `peers_from_control_nodes` True, then have all control nodes peer to it
if address has `peers_from_control_nodes` False, then remove this peer from all control nodes.

I do a bit of set() comparisons to make sure the add/remove is necessary, because I want to minimize the amount of times we call `schedule_write_receptor_config`

AlanCoding commented on awx/api/views/__init__.py:
```diff
@@ -371,13 +371,38 @@ def get_queryset(self):
+    search_fields = 'address'
+
+
+class InstanceReceptorAddressesList(ListCreateAPIView):
```
should this be a `SubListCreateAPIView`?

It looks like you're trying to create a list based on a reverse ForeignKey relationship. I don't love that when used for creation, but it's done here-and-there, and this is as valid of a case as any. This type of view needs the `parent_key` set (I assume `instance`).

AlanCoding commented on awx/api/views/__init__.py:
```diff
@@ -371,13 +371,38 @@ def get_queryset(self):
+        return super().post(request, *args, **kwargs)
+
+
+class ReceptorAddressesList(ListAPIView):
```
Clients have a hard time when a resource can be created in a sublist, but not in a global list. Since address->instance relationship is a ForeignKey, the contract is easy to manage. Just make sure `null=False` is set on the `instance` field, and the on_delete behavior is already good.

fosterseth committed:
Add API validation when creating ReceptorAddress

- websocket_path can only be set if protocol is ws
- is_internal must be False
- only 1 address per instance can have
peers_from_control_nodes set to True

Signed-off-by: Seth Foster <fosterbseth@gmail.com>

fosterseth committed:
Add validation when setting peers

- cannot peer to self
- cannot peer to instance that is already peered to self

Other changes:
- ReceptorAddress protocol field restricted to choices: tcp, ws, wss
- fix awx-manage list_instances when instance.last_seen is None
- InstanceLink make source and target unique together
- Add help text to the ReceptorAddress fields

Signed-off-by: Seth Foster <fosterbseth@gmail.com>

fosterseth committed:
Add functional and unit tests

Updated existing tests to support the
ReceptorAddress model

- cannot peer to self
- cannot peer to node that is already peered to me
- cannot peer to node more than once (via 2+ addresses)
- cannot set is_internal True

Other changes:
Change post save signal to only call
schedule_write_receptor_config() when an actual change is detected.

Make functional tests more robust by
checking for specific validation error in the
response.

I.e. instead of just checking for 400, just for 400
and that the error message corresponds to the
validation we are testing for.

Signed-off-by: Seth Foster <fosterbseth@gmail.com>

chrismeyersfsu commented on awx/main/models/ha.py:
```diff
@@ -509,10 +526,12 @@ def on_instance_saved(sender, instance, created=False, raw=False, **kwargs):
-            instance.peers.set(inst)
+        peers_address = ReceptorAddress.objects.filter(peers_from_control_nodes=True)
+        with disable_activity_stream():
+            instance.peers.add(*peers_address)
```
The uniqueconstraint on `<source, target>` ensures that the list of `ReceptorAddress` related to the instance is, effectively, a set. :+1: 

My worry was that this might add duplicates, but I've convinced myself that is not the case. :+1:

fosterseth replied:
yeah the unique constraint on "source" and "target" set is necessary to prevent duplicate peering

fosterseth committed:
Update awx_collection to support ReceptorAddress

- Add receptor_address module which allows
users to create addresses for instances

- Update awx_collection functional and integration
tests to support new peering design

Signed-off-by: Seth Foster <fosterbseth@gmail.com>

fosterseth committed:
Add search fields to views

Make receptoraddress list views
searchable by "address"

Other changes:
- Add help text to source and target of the
InstanceLink model

Signed-off-by: Seth Foster <fosterbseth@gmail.com>

fosterseth committed:
Remove unused variables and imports

Addresses flake8 and linting failures

Signed-off-by: Seth Foster <fosterbseth@gmail.com>

fosterseth committed:
Add choices to module protocol field

Signed-off-by: Seth Foster <fosterbseth@gmail.com>

fosterseth committed:
Fix inconsistent tab width

Signed-off-by: Seth Foster <fosterbseth@gmail.com>

fosterseth committed:
Fix proper indent to instance module

Signed-off-by: Seth Foster <fosterbseth@gmail.com>

fosterseth commented on awx/main/models/receptor_address.py:
```diff
@@ -0,0 +1,65 @@
+            )
+        ]
+
+    class Protocols(models.TextChoices):
```
pull out of ReceptorAddress and just call it ReceptorProtocols so that it can easily be used elsewhere

jbradberry commented on awx/api/serializers.py:
```diff
@@ -5356,24 +5357,67 @@ def validate(self, attrs):
+
+    def validate(self, attrs):
+        def get_field_from_model_or_attrs(fd):
+            return attrs.get(fd, self.instance and getattr(self.instance, fd) or None)
```
I recommend using the official Python trinary statement `A if B else C` instead.

jbradberry commented on awx/api/serializers.py:
```diff
@@ -5356,24 +5357,67 @@ def validate(self, attrs):
+
+        # an instance can only have one address with peers_from_control_nodes set to True
+        if peers_from_control_nodes:
+            for other_address in ReceptorAddress.objects.filter(instance=instance.id):
```
I'd use a more compact queryset statement instead of this explicit loop.  Something like

```
if ReceptorAddress.objects.filter(instance=instance.id, peers_from_control_nodes=True).exists():
    raise ValidationError(...)
```

jbradberry replied:
What happens if the user is editing an existing address instead of creating a new one?  Won't either of these versions of the validation logic count the existing one?

jbradberry replied:
Probably correctable using an `and` clause on line 5409.  I forget what the logic is exactly for checking if a serializer is getting an existing object, but it probably has to do with seeing if the `id` is available and set on some part of the data.

jbradberry commented on awx/api/serializers.py:
```diff
@@ -5356,24 +5357,67 @@ def validate(self, attrs):
+                    raise serializers.ValidationError(_("Only one address can set peers_from_control_nodes to True."))
+
+        # is_internal should be False
+        if attrs.get('is_internal') == True:
```
Why isn't this field read-only?

jbradberry commented on awx/api/serializers.py:
```diff
@@ -5481,41 +5526,43 @@ def check_peers_changed():
+        if self.instance and self.instance.receptor_addresses.filter(id__in=peers_ids).exists():
+            raise serializers.ValidationError(_("Instance cannot peer to its own address."))
+
+        if self.instance and self.instance.receptor_addresses.all().exists():
```
I wouldn't issue this queryset twice.  Drop it here and the one on the next line ought to be sufficient (you'll just get an empty set if `.exists()` would have returned `False` here).

jbradberry commented on awx/api/serializers.py:
```diff
@@ -5481,41 +5526,43 @@ def check_peers_changed():
+
+        # cannot peer to instance more than once
+        # compare length of set to original list to check for duplicates
+        peers_instances = [p.instance for p in attrs.get('peers', [])]
```
You should probably do this block before the previous one, since it doesn't involve querying the db just checking the user's input.

jbradberry commented on awx/main/management/commands/register_peers.py:
```diff
@@ -39,6 +39,12 @@ def handle(self, **options):
         if options['exact'] is not None and options['disconnect']:
             raise CommandError("The option --disconnect may not be used with --exact.")
 
+        # make sure peers and source instances only have one receptor address
```
???

dmzoneill committed:
UI Updates for receptor peering

fosterseth committed:
Add canonical receptor address

Creates a non-deletable address that acts as
the "main" address for this instance.

All other addresses for that instance must
be non-canonical.

When listener_port on an instance is set, automatically
create a canonical receptor address where:
  - address is hostname of instance
  - port is listener_port
  - canonical is True

Additionally, protocol field is added to instance to
denote the receptor listener protocol to use (ws, tcp).

The receptor config listener information is derived from
the listener_port and protocol information. Having a
canonical address that mirrors the listener_port ensures that
an address exists that matches the receptor config information.

Other changes:
- Add managed field to receptor address.
If managed is True, no fields on on this address can be edited
via the API.
If canonical is True, only the address cannot be edited.

- Add managed field to instance. If managed is True, users
cannot set node_state to deprovisioning (i.e. cannot delete node)

This change to our mechanism to prevent users from deleting
the mesh ingress hop node.

- Field is_internal is now renamed to k8s_routable

- Add reverse_peers on instance which is a list of instance IDs
that peer to this instance (via an address)

Signed-off-by: Seth Foster <fosterbseth@gmail.com>

dmzoneill committed:
Fix lint trailing whitespace

jbradberry commented on awx/main/migrations/0188_inbound_hop_nodes.py:
```diff
@@ -0,0 +1,106 @@
+# Generated by Django 4.2.6 on 2023-12-14 19:14
```
This migration has no data migration.  From what I see currently, we need:

- [ ] create a canonical `ReceptorAddress` when the listener is turned on, with copies of the port, address, and `peers_from_control_plane` (as an aside, does `ReceptorAddress` actually need this field?)
- [ ] existing `InstanceLink` entries need to migrate their `target` data

jbradberry commented on awx/main/models/receptor_address.py:
```diff
@@ -0,0 +1,59 @@
+    k8s_routable = models.BooleanField(help_text=_("If True, only routable inside of the Kubernetes cluster."), default=False)
+    canonical = models.BooleanField(help_text=_("If True, this address is the canonical address for the instance."), default=False)
+    managed = models.BooleanField(help_text=_("If True, this address is managed by the control plane."), default=False, editable=False)
+    peers_from_control_nodes = models.BooleanField(help_text=_("If True, control plane cluster nodes should automatically peer to it."), default=False)
```
Upon reflection, I think that this field should not exist.  `Instance.peers_from_control_nodes` ought to be sufficient, just follow the fk relationship from the instance to the canonical address.

dmzoneill committed:
Mesh UI support

- add endpoint
- delete endpoint (wip)
- associate
- disassociate

TheRealHaoLiu committed:
Fix provision instance not respecting protocol

dmzoneill committed:
Cleanup

- Remove peer selection on add and edit instance
- Added conconcial name and order comlums names same on endpoints and
  peers
- Other cleanup items
- rename name to instance name

CFSNM committed:
Comment unused dependency

CFSNM committed:
Adjust migration names and dependencies

jbradberry committed:
Join across the InstanceLink.target to the underlying Instance

to avoid having to do a lookup for every instance in the list for the
mesh visualizer endpoint.

fosterseth committed:
Remove CRUD for Receptor Addresses

Removes ability to directly create and delete
receptor addresses for a given node.

Instead, receptor addresses are created automatically
if listener_port is set on the Instance.

For example patching "hop" instance

with {"listener_port": 6667}

will create a canonical receptor address with port
6667.

Likewise, peers_from_control_nodes on the instance
sets the peers_from_control_nodes on the canonical
address (if listener port is also set).

protocol is a read-only field that simply reflects
the canonical address protocol.

Other Changes:

- rename k8s_routable to is_internal
- add protocol to ReceptorAddress
- remove peers_from_control_nodes and listener_port
from Instance model

Signed-off-by: Seth Foster <fosterbseth@gmail.com>

fosterseth committed:
Rename migration dependency

Signed-off-by: Seth Foster <fosterbseth@gmail.com>

fosterseth committed:
Make canonical field default to False

Signed-off-by: Seth Foster <fosterbseth@gmail.com>

fosterseth committed:
Remove receptor_address module from collection

After removing CRUD from receptor addresses, we need
to remove the module.

- remove receptor_address module
- Add listener_port to instance module
- Add peers_from_control_nodes to instance module

Signed-off-by: Seth Foster <fosterbseth@gmail.com>

TheRealHaoLiu committed:
Fix lint error, remove unused import

TheRealHaoLiu committed:
Fix ui-lint error

fosterseth committed:
Add canonical=True when creating ReceptorAddress in tests

Signed-off-by: Seth Foster <fosterbseth@gmail.com>

TheRealHaoLiu committed:
Update bootstrap_development.sh

fosterseth committed:
Ensure register_peers target is ReceptorAddress

command has 3 "targets", all of which should be a
ReceptorAddress object

- peers, disconnect, exact

Add logic to make sure each entry in those lists
are receptor addresses.

When creating InstanceLink objects, make sure
target is ReceptorAddress, not an Instance.

Signed-off-by: Seth Foster <fosterbseth@gmail.com>

fosterseth committed:
Remove unused warnings import

Signed-off-by: Seth Foster <fosterbseth@gmail.com>

fosterseth committed:
Add protocol to receptor address serializer

Signed-off-by: Seth Foster <fosterbseth@gmail.com>

TheRealHaoLiu committed:
Fix condition for creating receptor_address

If listener_port is not explicitly defined don't create a receptor_address

fosterseth committed:
Only create receptor address if port is defined

If a Instance endpoint is patched with
{"peers_from_control_nodes" True}

but a listener_port is not defined on the instance,
or is part of the patch payload, do not create a
receptor address.

Only update or create a receptor address if listener_port
is set, either in the payload or already on the instance.

Signed-off-by: Seth Foster <fosterbseth@gmail.com>

fosterseth committed:
Add migration to support InstanceLink changes

- Add forwards method to create a receptor address
for any existing Instance that has listener_port defined

- Add forwards method to modify each InstanceLink object
that changes target to the newly created receptor addresses

This migration was implemented as follows:

1. Add a target_new to InstanceLink which is a foreign key
to ReceptorAddress
2. create receptor addresses
3. link to these receptor addresses using the target_new field
4. rename target_new to target
5. drop listener_port and peers_from_control_nodes from Instance

Signed-off-by: Seth Foster <fosterbseth@gmail.com>

dmzoneill committed:
Updates for receptor reaslese to ui for protocol and is_managed

dmzoneill committed:
Fix remaning tests, removed unused code

dmzoneill committed:
InstanceAdd sends null for port_listener

fosterseth committed:
Add management command to remove address

Adds remove_receptor_address to delete a
receptor address from the database

Also, enforce that only 1 canonical address
can be added to an instance via
the add_receptor_address command.

Signed-off-by: Seth Foster <fosterbseth@gmail.com>

jbradberry committed:
Reconstitute migration file

Do the things as a proper 3-step process -- add new stuff, copy the data, drop the old.

fosterseth committed:
Make InstanceLink target non-nullable

InstanceLink target should not be null.

Should be safe to set to null=False, because we have
a custom RunPython method to explicitly set
target to a proper key.

Also, add new test to test_migrations
which ensures data integrity after migrating
the receptor address model changes.

Signed-off-by: Seth Foster <fosterbseth@gmail.com>

jbradberry committed:
Template the listener protocol into the receptor install bundle (#14792)

Previously we were hard-coding tcp, now we need to also support ws/wss.

jbradberry committed:
Support wss as ws-listener in the Receptor config

TheRealHaoLiu committed:
Update requirements.yml

dmzoneill committed:
Form hardening and node type exclusion

Disabled add/edit/remove for managed nodes
Tightened validation on Peers from control nodes

fosterseth committed:
Prevent modifying peers on managed node

Add validation to prevent any managed node
from modifying "peers" through the API

Peering from these nodes should be handled
by setting peers_from_control_nodes only.

Managed nodes are control nodes and
ingress hop nodes.

Signed-off-by: Seth Foster <fosterbseth@gmail.com>

fosterseth committed:
Require receptor collection 2.0.3

This release supports ingress hop node, and
is backwards compatible with previous AWX
version.

Signed-off-by: Seth Foster <fosterbseth@gmail.com>

fosterseth committed:
Peers_from_control_nodes requires listener port

Adds validation and a unit test to ensure:

- peers_from_control_nodes=True should fail if
listener_port is not set
- peers_from_control_nodes=False should be NOOP if
listener_port is not set

Signed-off-by: Seth Foster <fosterbseth@gmail.com>

jbradberry committed:
The listener port cannot be disabled when setting peers_from_control_nodes

If the port is explicitly set to null (causing any ReceptorAddress to
be deleted), then that's a validation error.

If the port is left off but a ReceptorAddress doesn't already exist,
we should not infer a port number and that is also a validation error.

jbradberry committed:
'managed' is a read-only field on InstanceSerializer

so we don't need complex logic to compare an incoming to existing value.

jbradberry committed:
Use Counter to find duplicate peer relationships

Gives a bit more readability.

jbradberry committed:
Placeholder FIXMEs for things of concern

jbradberry committed:
Use a select_related to build the peers queryset in the install bundle

Since the relationship is ReceptorAddress -> Instance,
prefetch_related isn't necessary.

jbradberry committed:
Make the peer validation more compact

and reuse information.

jbradberry committed:
Break out peer validation into its own method

jbradberry committed:
Write tests around the two special instance serializer fields

and all of the cases that they might be in.

fosterseth committed:
Test inspect_established_receptor_connections

Add functional test case for inspecting
established receptor connections.

InstanceLink starts in ADDING state, and should
move to ESTABLISHED state if the connection
is detected in the receptor status output.

Signed-off-by: Seth Foster <fosterbseth@gmail.com>

fosterseth committed:
If managed, cannot modify peers_from_control_nodes

Adds validation to prevent changing
peers_from_control_nodes if instance managed=True

Signed-off-by: Seth Foster <fosterbseth@gmail.com>

fosterseth committed:
Remove redundant tests

test_listener_port
test_peers_from_control_nodes
test_peers_from_control_nodes_without_listener_port

are covered in the following tests:

test_no_op
test_creates_canonical_address
test_deletes_canonical_address
test_updates_canonical_address
test_canonical_address_validation_error

Signed-off-by: Seth Foster <fosterbseth@gmail.com>

fosterseth committed:
Prevent duplicating instance links

In receptor address post-save method:
- Fixed detecting if address was missing
a link from control nodes
- Use InstanceLink create_or_update to prevent
adding duplicate InstanceLink source and target
peers

In instance serializer create_or_update,
delete receptor addresses first before doing
instance create or update. This ensures that we don't
trigger unnecessary post-save methods that might
attempt to manipulate receptor addresses that
will just be removed later.

Signed-off-by: Seth Foster <fosterbseth@gmail.com>

fosterseth committed:
Protocol blank if no canonical address

Make protocol be blank on instance if there
is no canonical address for this instance.

It was defaulting to "tcp" before.

Signed-off-by: Seth Foster <fosterbseth@gmail.com>

fosterseth committed:
Fix UI lint by running npm prettier

Signed-off-by: Seth Foster <fosterbseth@gmail.com>

fosterseth committed:
Disable health check button if managed

Also, update ui screen tests to expect
injecting "listener_port: null" if
listener_port is empty

Signed-off-by: Seth Foster <fosterbseth@gmail.com>

fosterseth committed:
InstanceLink unique constraint source and target

Prevent creating InstanceLinks with duplicate
source and target pairings.

Signed-off-by: Seth Foster <fosterbseth@gmail.com>

fosterseth committed:
UI rename Endpoints to Listener Addresses

Listener Addresses is a better name to
emphasize these are routable addresses to
reach a listener service on the node.

Also removed expand toggle on the listener
addresses list items, as the expanded mode
had no additional information.

Signed-off-by: Seth Foster <fosterbseth@gmail.com>
